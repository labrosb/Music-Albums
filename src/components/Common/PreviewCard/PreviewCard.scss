@import '../../../styles/colors.scss';
@import '../../../styles/mixins.scss';

.card-container {
  width: 200px;
  margin: 12px 2vw;
  background-color: color(background);
  cursor: pointer;
  transition: transform 200ms ease-in-out;
  position: relative;

  .card-image {
    width: 100%;
    margin: auto;
    border: 0.5px solid color(fade);
  }

  .card-favorite {
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 14px;
    color: color(background);
    border: 1px solid color(background);
    border-radius: 50%;
    height: 30px;
    width: 32px;
    opacity: 0;
    position: absolute;
    top: 10px;
    left: 10px;
    transition: opacity 250ms ease-in-out;
    z-index: -1;
  }
  .card-favorite.active {
    background-color: color(favorite);
    border: 1px solid color(favorite);
    opacity: 0.9;
    z-index: 1;
  }

  .inline-elements {
    display: flex;

    .card-rank {
      display: flex;
      align-items: center;
      justify-content: center;
      height: 24px;
      min-width: 24px;
      margin-top: 5px;
      font-size: 12px;
      font-weight: 500;
      color: color(text);
      background-color: color(special);
      border: 1px solid color(secondary);
      border-radius: 50%;
    }

    .card-details {
      box-sizing: border-box;
      padding-left: 8px;
      overflow: hidden;

      .card-detail {
        margin: 6px 0;
        text-overflow: ellipsis;
        overflow: hidden;
        white-space: nowrap;
      }
      .card-detail.primary {
        @include text(small, text);
      }
      .card-detail.secondary {
        @include text(tiny, text);
      }
    }
  }
}
.card-container:hover {
  transform: translate(4px, -12px) skewY(-4deg);
  box-shadow: 0px 0px 60px 0px rgba(0, 0, 0, 0.06);
  .card-favorite {
    opacity: 0.86;
    z-index: 1;
  }
}

@media (max-width: 480px) {
  .card-container {
    width: 43vw;
    margin: 12px 1vw;
  }
}
