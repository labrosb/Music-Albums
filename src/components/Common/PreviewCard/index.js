import React, { useCallback } from 'react';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart } from '@fortawesome/free-solid-svg-icons';
import './PreviewCard.scss';

// Preview Card Component used as list / Catalogue Item

function Preview({ toggleFavorite, isFavorite, album }) {
  const history = useHistory();
  const onCardClick = useCallback(() => {
    history.push({
      pathname: `/album-page/${album.id}`,
      state: album
    });
  }, [album, history]);

  const onFavoriteClick = useCallback(ev => {
    ev.stopPropagation();
    toggleFavorite(isFavorite, album);
  }, [isFavorite, album, toggleFavorite]);

  return (
    <div
      className="card-container"
      onClick={onCardClick}
    >
      <img
        className="card-image"
        src={album.thumbnail}
        alt={album.thumbSmall}
      />
      <div
        className={`card-favorite ${isFavorite ? 'active' : ''}`}
        onClick={onFavoriteClick}
      >
        <FontAwesomeIcon
          className="favorite-icon"
          icon={faHeart}
        />
      </div>
      <div className="inline-elements">
        <span className="card-rank">
          {album.rank || 'X'}
        </span>
        <div className="card-details">
          <p className="card-detail primary">
            {album.name}
          </p>
          <p className="card-detail secondary">
            {album.artist}
          </p>
        </div>
      </div>
    </div>
  );
}

Preview.propTypes = {
  isFavorite: PropTypes.bool,
  album: PropTypes.object,
  toggleFavorite: PropTypes.func
};

export default Preview;
