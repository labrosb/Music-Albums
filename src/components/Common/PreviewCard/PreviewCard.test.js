/* eslint-disable no-undef */
import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import PreviewCard from './index';

configure({ adapter: new Adapter() });

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn()
}));

const album = {
  artist: 'Kenny Chesney',
  category: 'Country',
  id: '1502163294',
  link: 'https://music.apple.com/us/album/here-and-now/1502163294?uo=2',
  name: 'Here and Now',
  price: '$10.99',
  rank: 1,
  releaseDate: 'May 1, 2020',
  rights: { label: 'â„— 2020 Blue Chair Records, LLC under exclusive license to Warner Music Nashville LLC.' },
  thumbSmall: 'https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/3c/a4/90/3ca4901d-b744-411d-246e-02f78aebf8bb/093624892984.jpg/55x55bb.png',
  thumbnail: 'https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/3c/a4/90/3ca4901d-b744-411d-246e-02f78aebf8bb/093624892984.jpg/500x500bb.png'
};

let wrapper;
beforeEach(() => {
  wrapper = shallow(<PreviewCard isFavorite {...{ album }} />);
});

describe('PreviewCard component', () => {
  it('Should render image correctly', () => {
    const imageSrc = wrapper.find('.card-image').prop('src');

    expect(imageSrc).toEqual(album.thumbnail);
  });

  it('Should render ranking correctly', () => {
    const rank = wrapper.find('.card-rank').text();

    expect(rank).toEqual(album.rank.toString());
  });

  it('Should render album name correctly', () => {
    const name = wrapper.find('.card-detail').first().text();

    expect(name).toEqual(album.name);
  });

  it('Should render artist name correctly', () => {
    const artist = wrapper.find('.card-detail').at(1).text();

    expect(artist).toEqual(album.artist);
  });

  it('Should have active favorite button', () => {
    const favorite = wrapper.find('.card-favorite');

    expect((favorite).hasClass('active')).toBe(true);
  });

  it('Should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
