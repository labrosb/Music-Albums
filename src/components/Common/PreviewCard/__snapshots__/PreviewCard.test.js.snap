// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewCard component Matches snapshot 1`] = `
<div
  className="card-container"
  onClick={[Function]}
>
  <img
    alt="https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/3c/a4/90/3ca4901d-b744-411d-246e-02f78aebf8bb/093624892984.jpg/55x55bb.png"
    className="card-image"
    src="https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/3c/a4/90/3ca4901d-b744-411d-246e-02f78aebf8bb/093624892984.jpg/500x500bb.png"
  />
  <div
    className="card-favorite active"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className="favorite-icon"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f004",
            "M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z",
          ],
          "iconName": "heart",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
  <div
    className="inline-elements"
  >
    <span
      className="card-rank"
    >
      1
    </span>
    <div
      className="card-details"
    >
      <p
        className="card-detail primary"
      >
        Here and Now
      </p>
      <p
        className="card-detail secondary"
      >
        Kenny Chesney
      </p>
    </div>
  </div>
</div>
`;

exports[`PreviewCard component Should match snapshot 1`] = `
<div
  className="card-container"
  onClick={[Function]}
>
  <img
    alt="https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/3c/a4/90/3ca4901d-b744-411d-246e-02f78aebf8bb/093624892984.jpg/55x55bb.png"
    className="card-image"
    src="https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/3c/a4/90/3ca4901d-b744-411d-246e-02f78aebf8bb/093624892984.jpg/500x500bb.png"
  />
  <div
    className="card-favorite active"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className="favorite-icon"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f004",
            "M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z",
          ],
          "iconName": "heart",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
  <div
    className="inline-elements"
  >
    <span
      className="card-rank"
    >
      1
    </span>
    <div
      className="card-details"
    >
      <p
        className="card-detail primary"
      >
        Here and Now
      </p>
      <p
        className="card-detail secondary"
      >
        Kenny Chesney
      </p>
    </div>
  </div>
</div>
`;
